ansible/roles/test/files/ptftests/advanced-reboot.py

```diff
@@ -957,6 +1043,163 @@ def wait_until_cpu_port_up(self):
                 break
             time.sleep(self.TIMEOUT)
 
+    def send_in_background(self, packets_list = None, interval = None):
+        """
+        This method sends predefined list of packets with predefined interval.
+        """
+        if not interval:
+            interval = self.send_interval
+        if not packets_list:
+            packets_list = self.packets_list
+        sender_start = datetime.datetime.now()
+        self.log("Sender started at %s" % str(sender_start))
+        for entry in packets_list:
+            time.sleep(interval)
+            testutils.send_packet(self, *entry)
+        self.log("Sender has been running for %s" % str(datetime.datetime.now() - sender_start))
+
+    def sniff_in_background(self, wait = None):
+        """
+        This function listens on all ports, in both directions, for the UDP src=1234 dst=5000 packets, until timeout.
+        Once found, all packets are dumped to local pcap file,
+        and all packets are saved to self.packets as scapy type.
+        """
+        if not wait:
+            wait = self.time_to_listen + 30
+        sniffer_start = datetime.datetime.now()
+        self.log("Sniffer started at %s" % str(sniffer_start))
+        filename = '/tmp/capture.pcap'
+        sniff_filter = "udp and udp dst port 5000 and udp src port 1234 and not icmp"
+        self.packets = scapyall.sniff(timeout = wait, filter = sniff_filter)
+        self.log("Sniffer has been running for %s" % str(datetime.datetime.now() - sniffer_start))
+        if self.packets:
+            scapyall.wrpcap(filename, self.packets)
+            self.log("Pcap file dumped to %s" % filename)
+        else:
+            self.log("Pcap file is empty.")
+
+    def send_and_sniff(self):
+        """
+        This method starts two background threads in parallel:
+        one for sending, another for collecting the sent packets.
+        """
+        self.sender_thr = threading.Thread(target = self.send_in_background)
+        self.sniff_thr = threading.Thread(target = self.sniff_in_background)
+        self.sniff_thr.start()
+        time.sleep(1)           # Let the listener initialize completely.
```

Comment:
probably it's better to use some sync. primitive to be sure you run your sender, when your sniffer is ready.

Reply:
+1

Reply:
I tried to sync sender and sniffer by threading Event,
but it didnt help: python actually makes sender to wait for sniffer,
but scapy.sniff is slow  and very unstable on start (it misses packets), and starts stable capturing actually ~1s later after python called it.

I dont know how to get notification from scapy.sniff that it has really started capturing, except the explicit time.wait(1).